File: events/2015/20/solution.go
package main

import (
	"encoding/json"
	"fmt"
	"math"
	"os"
	"strconv"
	"time"
	"yod_aoc/utils/go/aoc_utils"
)

func elvesWhoVisited(house int) []int {
	elves := []int{1, house}

	house_sqrt := int(math.Floor(math.Sqrt(float64(house))))

	for i := 2; i < house_sqrt; i++ {
		if house%i == 0 {
			elves = append(elves, i)

			house_i := int(math.Floor(float64(house / i)))

			if i != house_i {
				elves = append(elves, house_i)
			}
		}
	}
	return elves
}

func elvesWhoVisitedLimited(house int, limit int) []int {
	elves := []int{}

	house_sqrt := int(math.Floor(math.Sqrt(float64(house))))

	if limit >= house {
		elves = append(elves, 1)
	}

	for i := 2; i < house_sqrt; i++ {
		if house%i == 0 {
			if i*limit >= house {
				elves = append(elves, i)
			}

			house_i := int(math.Floor(float64(house / i)))

			if i != house_i && house_i*limit >= house {
				elves = append(elves, house_i)
			}
		}
	}

	elves = append(elves, house)

	return elves
}

func sumArr(arr []int) int {
	sum := 0

	for _, v := range arr {
		sum += v
	}

	return sum
}

// ----------------------------------------------------------------------------------------------------
// | Part 1
// ----------------------------------------------------------------------------------------------------
func part1(input []string) int {
	target, targetErr := strconv.Atoi(input[0])
	if targetErr != nil {
		panic("Unable to get Target Value from Input")
	}
	lastHouse := 0

	for true {
		elves := elvesWhoVisited(lastHouse + 1)
		sum := sumArr(elves) * 10
		if sum >= target {
			return lastHouse + 1
		}
		lastHouse += 1
	}

	return -1
}

// ----------------------------------------------------------------------------------------------------
// | Part 2
// ----------------------------------------------------------------------------------------------------
func part2(input []string) int {
	target, targetErr := strconv.Atoi(input[0])
	if targetErr != nil {
		panic("Unable to get Target Value from Input")
	}
	lastHouse := 0

	for true {
		elves := elvesWhoVisitedLimited(lastHouse+1, 50)
		sum := sumArr(elves) * 11
		if sum >= target {
			return lastHouse + 1
		}
		lastHouse += 1
	}

	return -1
}

// ----------------------------------------------------------------------------------------------------
// | Main Function
// ----------------------------------------------------------------------------------------------------
func main() {
	// Read the arguments
	args := os.Args[1:]
	var fileName string
	if len(args) > 0 {
		fileName = args[0]
	} else {
		fileName = "input.txt"
	}
	// Start the timer!
	startTime := time.Now()

	// Read the input to an array of strings
	path := aoc_utils.GetFilePath(2015, 20, fileName)
	content, err := aoc_utils.ReadInputAsArray(path)
	if err != nil {
		fmt.Println("Error reading input:", err)
		return
	}

	// Run the Parts
	p1Result := aoc_utils.MeasurePerformance(func() interface{} {
		return part1(content)
	})

	p2Result := aoc_utils.MeasurePerformance(func() interface{} {
		return part2(content)
	})

	// End the Timer!
	endTime := time.Now()
	duration := endTime.Sub(startTime)

	resultsForDay := aoc_utils.AOCDayResults{
		Year:     2015,
		Day:      20,
		Part1:    p1Result,
		Part2:    p2Result,
		Duration: duration.Nanoseconds(),
		Timestamp: aoc_utils.AOCTimestamp{
			Start: startTime.Format(time.RFC3339),
			End:   endTime.Format(time.RFC3339),
		},
	}
	jsonResults, _ := json.Marshal(resultsForDay)
	fmt.Println(string(jsonResults))
}

File: events/2015/20/solution.rs
use aoc_utils::read_input_as_array;
use aoc_utils::measure_performance;
use aoc_utils::AOCDayResults;
use aoc_utils::AOCTimestamp;
use chrono::{DateTime, Utc};

fn elves_who_visited(house: i32) -> Vec<i32> {
  let mut elves: Vec<i32> = Vec::new();
  elves.push(1);
  elves.push(house);

  let house_sqrt = ((house as f64).sqrt().floor()) as i32;

  for i in 2..=house_sqrt {
    if house % i == 0 {
      elves.push(i);

      if i != house/i {
        elves.push(house/i);
      }
    }
  }
  elves
}

fn elves_who_visited_limited(house: i32, limit: i32) -> Vec<i32> {
  let mut elves: Vec<i32> = Vec::new();
  
  if limit >= house {
    elves.push(1);
  }

  let house_sqrt = ((house as f64).sqrt().floor()) as i32;

  for i in 2..=house_sqrt{
    if house % i == 0 {
      if i * limit >= house {
        elves.push(i);
      }

      let paired = house/i;
      if i != paired && paired * limit >= house {
        elves.push(paired);
      }
    }
  }

  elves.push(house);
  
  elves
}

// ----------------------------------------------------------------------------------------------------
// | Part 1
// ----------------------------------------------------------------------------------------------------
fn part1(input: &Vec<String>) -> i32 {
  let target = input[0].parse::<i32>().unwrap_or(0);
  let mut last_house: i32 = 0;

  loop {
    let elves = elves_who_visited(last_house + 1);
    let sum = elves.into_iter().reduce(|acc, e| acc + e).unwrap_or(0) * 10;
    if sum >= target {
      return last_house + 1;
    }
    last_house += 1;
  }
}

// ----------------------------------------------------------------------------------------------------
// | Part 2
// ----------------------------------------------------------------------------------------------------
fn part2(input: &Vec<String>) -> i32 {
  let target = input[0].parse::<i32>().unwrap_or(0);
  let mut last_house: i32 = 0;

  loop {
    let elves = elves_who_visited_limited(last_house + 1, 50);
    let sum = elves.into_iter().reduce(|acc, e| acc + e).unwrap_or(0) * 11;
    if sum >= target {
      return last_house + 1;
    }
    last_house += 1;
  }
}

// ----------------------------------------------------------------------------------------------------
// | Main Function
// ----------------------------------------------------------------------------------------------------
fn main() {
  // Read the arguments
  let args = std::env::args().collect::<Vec<String>>();
  let default_file_name = String::from("input.txt");
  let file_name = args.get(1).unwrap_or(&default_file_name);

  // Start the timer
  let start_timestamp = std::time::SystemTime::now();
  let start_time = std::time::Instant::now();
  
  // Read the input
  let input = read_input_as_array(file_name).unwrap();

  // Run the Parts
  let p1_result = measure_performance(part1, &input);
  let p2_result = measure_performance(part2, &input);

  // Stop the timer
  let end_time = std::time::Instant::now();
  let end_timestamp = std::time::SystemTime::now();
  let duration = end_time.duration_since(start_time).as_nanos();

  let results = AOCDayResults {
    year: 2015,
    day: 20,
    part1: p1_result,
    part2: p2_result,
    duration: duration,
    timestamp: AOCTimestamp {
      start: DateTime::<Utc>::from(start_timestamp).to_rfc3339(),
      end: DateTime::<Utc>::from(end_timestamp).to_rfc3339(),
    },
  };

  let json_results = serde_json::to_string(&results).unwrap();
  println!("{}", json_results);
}